### DISCUSSION TITLE
Restack current branch, and follow.

### DISCUSSION THREAD

**Discussion by kevincox** [arxanas/git-branchless#228] [Ideas]
I'm new to git-branchless so maybe I missed an option. But in general I only want to restack the current commit/branch. I don't want to worry about other branches because solving conflicts may be a waste of time as more conflicts might appear as the parent gets modified. For the workflow I want I want to check out a commit or branch, then restack it, and be on the same branch with the new restacked commit.

Something like this seems reasonable:

```
git checkout my-branch
git restack HEAD
```

But at this point I am left on the old commit, separated from the branch. Would it be reasonable that if a commit that gets restacked is checked out the checkout follows to the new commit (and branch)?

This problem isn't fundamental. It can be solved just by checking out the branch again. But it is annoying so I thought I would open a discussion about the UX. I see no real reason why someone would want to be left on the old branch.

**Comment by arxanas** [comment]
Welcome @kevincox!

What version of git-branchless are you using? In recent versions, `git restack` should decline to restack any stacks which don't merge cleanly, which can work around a lot of the friction. (You have to pass `-m`/`--merge` to force merge conflict resolution.)

That being said, I think there's room to improve the interface for `git restack`, like maybe adding a `-a`/`--all` option to restack everything, and only restacking the current commit stack by default. I think the problem with that was that some commands leave you with abandoned commits that aren't descendants of the current commit, such as if you run `git rebase` and modify an ancestor commit to another sub-branch of the same stack.

In general, for `git restack`/`git move`, the checkout should follow the commit. If you have a specific repro case where it doesn't do that, then file an issue and I can look into it. The exception is if the `git checkout` fails after the operation, in which case the checkout of course wouldn't follow the rewritten commit. (If you install the latest version from source, @bcongdon committed #187 to improve the messaging in case of checkout failure.)

But to back up a moment, I don't think I understand how you're using `git restack`. Normally, I run `git restack` *after* an operation abandoned a commit, such as `git commit --amend` with descendant commits. But it sounds like you're using `git restack` *before* you want to start some work. What are you using it for in that case? Do you often have abandoned commits, and you defer cleaning them up until later?

If you're trying to rebase ("restack") the current commit onto the latest main branch, then `git restack` actually doesn't do that — you probably want an invocation like `git move -d main`. (I am working on an implementation of `git sync` to make this a little easier, and I also need to write documentation for handling this workflow, since it's pretty common.)

**Reply by kevincox** [reply]
I'm using 0.3.7.

> git restack should decline to restack any stacks which don't merge cleanly

Interesting. If this is a fairly seamless UX then I think just calling `git restack` should be good for my use case. I don't generally mind if extra branches get updated. I just don't want to deal with conflicts on branches that I am not currently focused on.

> git restack/git move, the checkout should follow the commit. If you have a specific repro case where it doesn't do that, then file an issue and I can look into it.

Sounds like I found a bug. The logs are gone now but what I did was pretty simple. I just checked out a branch that had an outdated parent, then ran `git restack` and I was no longer on that branch. My checkout was now dangling at the now outdated commit with no branch. This was quite confusing to me because `git sl` was showing angry symbols about the current commit and I couldn't figure why this old commit was still being shown until I realized that it was still checked out. Switching back to the branch (on the restacked commit) made everything go back to normal.

If I hit it again I'll open a bug with a more exact log of what happened. What type of info would you like to see in that case?

> But it sounds like you're using git restack before you want to start some work.

That's right. I am working on a tree of "patches" and I use branches to track each logical patch, basically what I would send out as a unit of review (it may be one commit, or multiple commits. but I would send them out together). What I do is whenever I watch to switch from one patch to another I will check it out, then if it is out of date in an important way I will restack it. For example maybe I made some API changes to the parent and want to update the child. Or maybe the parent was accepted so I want to restack onto the final commit. If only minor changes occurred such as formatting or renaming locals I won't bother restacking. (Although if restacking is quick and seamless I probably don't mind doing it occasionally, but most of the time I see no reason to especially if it risks merge conflicts). Basically once I start working on the child I will often "freeze" the parent to keep things simple then only restack when I need a bugfix/feature or I am finishing review and want to update the child to reflect the more final version.

So yes, the way restack works is backwards to me. I generally am not interesting in updating all children eagerly. I only update them when I want to continue work or finalize the review. After all what is the point of restacking these commits if I don't do anything with them?

> like maybe adding a -a/--all option to restack everything, and only restacking the current commit stack by default

This would make more sense to me but is still backwards to my thinking as I described earlier. The thing is that when I am working on a patch I don't really want to keep in my mind "Did I do something that requires restacking?" and "Does this commit have any children?". This is part of the reason why in my mind I would want to specify what commit to restack, instead of the parent who's children should be restacked. What works for me is seeing that a commit is based on an outdated parent in smartlog and restacking it. I don't have to keep any state in my head, I just fix the problem when it becomes relevant to me. Another way of thinking about this is when I am working on the parent I don't want to worry about the children. I'll fix up the children when I get back around to working on them.

That being said it still comes back to I don't want to deal with more problems than necessary right now. So if I want to work on PatchA and need to restack it I don't really care how many other things are restacked as long as it is invisible to me. But I definitely don't want to be bothered by other changes. I can see some edge cases where it could matter (for example you have a hot build cache on the last version and restacking would require a clean build) but for the most part restacking extra commits isn't too bad (and can be convenient if it saves manual restacks later).

> I am working on an implementation of git sync to make this a little easier

This would be nice. `restack` and `rebase` are definitely related operations. It would be nice if there was a generic `sync` that would rebase both from outdated commits and outdated branches. Both cases in my mind are "my parent is no longer the latest version, please fix that".

...anyways. This is why I opened this discussion. I really like the restacking concept allowing me to have multiple dependent patches locally while gracefully handing history rewrites. However there is clearly a mismatch between the current UX and how I think of it. My background with a similar workflow is using `hg` at Google so that may have coloured my way of thinking. It is entirely reasonable for you to say that this just isn't how git-branchless works and it won't change. I can slowly adapt or write some quick scripts to bridge the gap. But hopefully it is a useful insight to how some others are thinking about this type of workflow.

**Reply by arxanas** [reply]
> If I hit it again I'll open a bug with a more exact log of what happened. What type of info would you like to see in that case?

For this kind of bug report, it would be good to report which commands you ran, and also run `git undo` and copy the last few pages of relevant events (and smartlog structure, if relevant), and redact any sensitive information. There's an example of someone pasting the recent events at  https://github.com/arxanas/git-branchless/issues/209. The issue #214 tracks making this automated.

There is an unusual case at https://github.com/arxanas/git-branchless/issues/160 which might or might not be relevant in this case, where there are angry symbols in the smartlog that don't seem to go away when you expect them to.

> If this is a fairly seamless UX then I think just calling git restack should be good for my use case. I don't generally mind if extra branches get updated. I just don't want to deal with conflicts on branches that I am not currently focused on.

Yep, `git restack` and `git move` should do this seamlessly as of a recent version. I also hate not knowing if an operation will introduce merge conflicts.

> What I do is whenever I watch to switch from one patch to another I will check it out, then if it is out of date in an important way I will restack it. 

I see. Thanks for clarifying your workflow. Here's how I deal with these problems:

* Rather than amend a given patch (I assume that's what you're doing?), I always check it out and start making new WIP patches on top of it, with messages like `temp: fix foo`, etc. Doing so never causes the descendant commits to be abandoned, since it's a new line of work.
  * One ergonomic problem with always amending the commit in-place is that git-branchless doesn't have a good system for viewing the history of that commit over the time axis. You can view old versions in `git undo` and manually diff them, but it's kind of a pain.
  * Another ergonomic concern is that it's hard to see what the restack target for a given abandoned is. I suppose it says "rewritten as XXX" above it, but I don't generally spend time verifying that it's the commit I think it is.
  * Plus, you can do more speculative branching changes depending on the further commits you make. I have to write up the article [Workflow: Divergent development](https://github.com/arxanas/git-branchless/wiki/Workflow:-divergent-development) to detail a little more about it.
  * This pretty much amounts to strictly more information than you would have by amending in-place.
* When I want to confirm the changes to a patch, I collapse them together with `git rebase -i`. This causes other descendants to be abandoned, so I run `git restack` then. But there's no fundamental reason why that operation can't be delayed, like you do.
  * Alternatively, I sometimes `git move` the descendant branch directly onto the WIP patches, in case I don't want to e.g. upload the changes to the code review tool immediately.

I think your workflow should be supported fine (excepting whatever bug you saw), but I just happen not to do it that way. If you encounter additional issues, please continue to report them and I'll see what I can do.

> It would be nice if there was a generic sync that would rebase both from outdated commits and outdated branches. Both cases in my mind are "my parent is no longer the latest version, please fix that".

Oh, interesting — my conception of `git sync` was just to rebase all commit stacks on top of the main branch. From what I've seen, I think many users conceive of two distinct operations: "sync my work with the latest upstream main branch", and "sync my work with respect to itself." We'll have to see if it makes more sense to keep `git sync` and `git restack` separate and have the user make their own alias, or to add an additional flag to `git sync`.

**Reply by arxanas** [reply]
I would also suggest you take a look at https://github.com/martinvonz/jj/ (cc @martinvonz), the reason being that its workflow helps you delay conflict resolution more seamlessly. Commit evolution doesn't exist in jj (anymore); instead, descendant commits are always rebased on top of changed commits, and conflicts are stored in the commits themselves. So you never have to worry about resolving merge conflicts until you're ready to do so; they just show up as conflicting in the graph view. In some cases, the conflicts in descendant commits can even resolve themselves if you make the appropriate changes to its ancestor commits.

There might be a distant future where conflicts are stored inside commits in git-branchless as well, or at least *appear* to be in the smartlog, but it's a lot harder to integrate it with the surrounding Git tooling, which has no concept of conflicts in commits.

**Reply by arxanas** [reply]
> What I do is whenever I watch to switch from one patch to another I will check it out, then if it is out of date in an important way I will restack it.

How large is the repository you're working with? It occurs to me that I usually `restack`/`move` and then check out, in that order, because on large repositories, the checkout may take a long time, so it's best to minimize the number of checkouts. (It may also improve build artifact retention.) So often I instead write `git move -s <commit>` to move another stack of commits here, and then run `git next` to check out the first of those commits.

**Reply by arxanas** [reply]
I wrote up my general workflow in more detail at https://github.com/arxanas/git-branchless/wiki/Workflow:-divergent-development. But even with that being written, there's nothing wrong with a delayed-restack approach, like you use.

**Reply by kevincox** [reply]
> How large is the repository you're working with?

Not large. Checkout times are negligible.

**Reply by kevincox** [reply]
> Rather than amend a given patch (I assume that's what you're doing?), I always check it out and start making new WIP patches on top of it, with messages like temp: fix foo, etc. Doing so never causes the descendant commits to be abandoned, since it's a new line of work.

That's an interesting approach. I think the downside here is that often I want to send the parent for review after these fixes and so I need to rebase them into one to publish for review. So at least in the workflows I see to push for review as a rebased set I need to abandon the commits the children reference.

> Another ergonomic concern is that it's hard to see what the restack target for a given abandoned is. I suppose it says "rewritten as XXX" above it, but I don't generally spend time verifying that it's the commit I think it is.

I don't think this is a major problem for me because I generally don't care. I know it is some older version. If I think it might be relevant I would probably just rebase to be sure. But most of the time the exact parent isn't too relevant for me continuing development.

> I think your workflow should be supported fine (excepting whatever bug you saw), but I just happen not to do it that way. If you encounter additional issues, please continue to report them and I'll see what I can do.

Yeah, I think it should. It would be nice to delay the restack but not critical for my current projects. Even the bug is a minor inconvenience as I could wrap it into a script that checked out the right branch again. Thanks for the tool, it is really improving my workflow.

**Reply by JeffFaer** [reply]
At the risk of necroing this...

> That being said, I think there's room to improve the interface for git restack, like maybe adding a -a/--all option to restack everything, and only restacking the current commit stack by default

I think I'd also want to see that type of behavior with `git sync`, where by default it only affects the current commit and its parents, and maybe also its descendants. I often have a lot of things going on in parallel, and sometimes I abandon stuff for a while, and I'd rather not have to figure out any merge conflicts (whether they be real file-level merge conflicts or logical merge conflicts where the code will suddenly stop working until it's refactored to work with whatever changes were introduced on main) until I have some spare time to actually dedicate to the task

**Reply by arxanas** [reply]
@JeffreyFalgout Originally, `git sync` only synced the current commit stack, but, in that case, there wasn't much difference between doing that and running `git move -s 'stack()' -d 'main()'`. To support your workflow, I recommend you avoid using `git sync` directly and either run `git sync 'stack()'` or `git move` in some form.

(Honestly, the revset argument to `git sync` is not that useful, but it's there for consistency.)

**Comment by NonLogicalDev** [comment]
@arxanas I think it would be invaluable to be able to delay conflict resolution for later. This is one of the big things that I think Stacked Git (https://github.com/stacked-git/stgit) has gotten right.

I can have upwards of 15 different features that are based on top of `main` branch. Some are experiments that I work off and on. With stacked git I have the ability to restack those changes only when I feel like working on them.

I think Git Branchless has a lot of merits, for one I really like how it handles commit trees. But restacking operation that locks you in a stateful workflow is holding me back from being able to use it effectively.

It feels like Git Branchless could have either `git goto --restack` or `git restack --recursive HEAD` (restack all descendants of this commit) to update the commit you are moving to. Otherwise, if you have a lot of stacked changes you might be stuck in a "rebase" workflow for quite a while, and that gets us right back to `git rebase` issues we are trying to move away from.

Should you have to context switch to a different branch to work on something else or work on a hotfix, you either have to power through conflict resolutions or abort entirely. A more surgical and essentially stateless restack operation (like in Stacked Git) would be a huge boon.

**Reply by arxanas** [reply]
> I can have upwards of 15 different features that are based on top of main branch. Some are experiments that I work off and on. With stacked git I have the ability to restack those changes only when I feel like working on them.

Do you mean that the features are all stacked on each other, or that you have 15 different branches whose common ancestor is `main`?

If the former, there isn't currently a way to explicitly "cut out" a single commit without attempting to restack its descendants, but you could use this workaround instead: first, run `git commit --amend` (optionally with `--no-edit`) to cut out the current commit and leave its descendants abandoned, and then optionally use `git move` or whatever to move it to its intended destination.

> But restacking operation that locks you in a stateful workflow is holding me back from being able to use it effectively.

Which state are you talking about specifically? That commit rewrites are tracked, so that `restack` knows where to move commits? If so, you can opt to either ignore the "rewritten" annotations in the smartlog, or copy commits instead of moving them with `git cherry-pick`.

> It feels like Git Branchless could have either git goto --restack or git restack --recursive HEAD (restack all descendants of this commit) to update the commit you are moving to.

`git restack` accepts a commit as an option indicating what to restack, although it represents "The IDs of the abandoned commits whose descendants should be restacked. If not provided, all abandoned commits are restacked", so you'd have to specify the old commit which has been abandoned. (This could be improved if we think specifying the target commit is better). Otherwise, if you want to manually specify which commits to move, you can run e.g. `git move -s 'children(<target>)' -d <dest>` to rebase them manually.

I also recommend you take a look at [Jujutsu](https://github.com/martinvonz/jj/), since it can defer conflict resolution arbitrarily, because conflicts are stored in the commits themselves, and therefore rebases always succeed.

**Reply by GiorgosAlexakis** [reply]
`git rerere` is your friend

**Comment by arxanas** [comment]
By the way, these days, you can run e.g. `git restack 'stack()'` to restack only rewritten commits in the current commit stack; see https://github.com/arxanas/git-branchless/wiki/Reference:-Revsets
